### ISSUE TITLE
[CA performance] Pod condition reseting takes too much time.

### ISSUE THREAD

**Issue by mwielgus** [kubernetes/autoscaler#187]
With lots of pending pods we try to reset all of their unschedulable pod conditions every time a new node is added. That leads to performance problems (updates are synchronous) and large api server traffic.

cc: @aleksandra-malinowska @bskiba @MaciekPytel

**Comment by MaciekPytel** [comment]
tl;dr I've run some test and went through code and I think it's safe to just completely remove this logic. I'll create a PR tomorrow after I run all e2e tests.

===

Currently logic related to PodScheduled status condition filters out the pods that have been unschedulable before a new node was added and haven't been looked at by scheduler since it was added. It prevents us from triggering scale-up for those pods, as they may yet fit onto new node(s). I've run some tests and it happens pretty much always after we add a new node for a few loops, as it takes time for scheduler to try and fit new pods.

However, just after this is done we filter out all the pods that can be scheduled on existing nodes according to our simulator (FilterOutSchedulable function). I've run some manual test and this check normally catches and filters out all the pods that would be filtered out by PodScheduled condition filter.

So basically in normal usage removing this logic doesn't change anything as the pods it filters out from scale-up considerations would be filtered out by other, cheaper tests. 

There are only 2 situations where removing the PodScheduled related logic changes CA behavior:
1. Scheduler haven't processed the pod yet but it's unschedulable anyway (both in reality and in CA simulation). In this case we will trigger the scale-up earlier than before. Cool.
2.  Scheduler haven't processed the pod yet and we think it's unschedulable. Except we're wrong and the scheduler manages to fit it in after we've triggered an unnecessary scale-up. Bummer.

In reality, however, either of this situations is very unlikely to happen. Assuming the new node was added by CA (which is expected in autoscaled cluster), we would have run scale-up simulations while it was coming up. Since the node haven't existed yet it would not cause the pod to be filtered out by PodScheduled checks, instead we wouldn't trigger scale-up for it as because our simulation told us it would fit on the upcoming node. If our simulation was wrong we would trigger the unnecessary scale-up long before checking PodScheduled condition could safe us from it.

That being said I can come up with a few scenarios where it could help. They all assume that our simulation differs from actual scheduler though. 
If that's the case and additionally we overestimated the size of the node in scale-from-0. We may have believed the pod will fit while the node was upcoming, than figured out it actually won't after slightly-smaller-than-predicted node showed up, only to find out we were wrong when scheduler actually managed to fit the pod on it after all. Or we could have a race conditions between scheduler, CA and manually added node showing up instead of bad scale-from-0 node template.

Either way this is pretty unlikely and if our scheduler simulation is wrong we arguably have bigger problems than adding an extra node in some unlikely edge case scenario. 

===

If we remove this check setting `--verify-unschedulable-pods=false` will do very bad things though. Currently it just disables an additional filter for strange edge cases. But once we remove PodScheduled based filtering setting it will cause us to do multiple scale-ups almost every time.

Given that this flag is not part of any default or suggested configuration, it's completely undocumented, I never heard about anyone actually changing it and changing it will completely break CA I propose we completely remove this flag. 

===

Finally I think there are some more issues around filtering out unschedulable pods before we start scale-up. 
1. If CA simulation incorrectly says a pod can be scheduled, but the scheduler disagrees we will forever filter that pod out in FilterOutSchedulable and never trigger scale-up for it. This is correct behavior (if our simulator incorrectly predicts where this pod can schedule we don't know if a scale-up will really help it). Except we will never let the user know what's happening. I think we should create an event or something if a pod is filtered out for a long time.
2. FilterOutSchedulable checks if every pod can fit somewhere on existing node. But it doesn't update the simulation state. So if we have 100 pods with 100 millicore request and 150 millicores free on a single node we will not trigger a scale-up before the scheduler schedules one of those pods, so the other don't fit anymore. We should instead use binpacking, similar to what we do for upcoming nodes.

WDYT @mwielgus? If you agree I'll create separate issues to track those problems.

cc: @aleksandra-malinowska @bskiba

**Comment by MaciekPytel** [comment]
Noticed one more difference introduced by #189: after a new node is added, but scheduler haven't managed to put previously unschedulable pods on it we will now set schedulablePodsPresent to true, blocking scale-down until the scheduler deals with pending pods. I think this is actually an improvement - previous approach was inconsistent.

**Comment by mwielgus** [comment]
@MaciekPytel On vacation, will respond on Monday.

**Comment by mwielgus** [comment]
Main thing - ok. 
Extra stuff:
1. No. It will add yet another level of complexity. 
2. Ok, if there is some other feature that would also benefit from this (like slack)

**Comment by mwielgus** [comment]
Fixed with #189
